### PR TITLE
make equals() nothrow

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: compiler/src/dmd/func.d [hunk 1]
```diff
@@ -2033,6 +2034,7 @@ extern (C++) class FuncDeclaration : Declaration
                 return 0;
             }
         });
+        catch (Throwable) assert(0);
```
Comment (adamdruppe) [inline compiler/src/dmd/func.d:2037]: it should only catch Exception

File: compiler/src/dmd/mtype.d [hunk 2]
```diff
@@ -2386,13 +2386,15 @@ extern (C++) abstract class Type : ASTNode
...
+    inout(Type) toHeadMutable() inout nothrow
     {
         if (!mod)
             return this;
         Type unqualThis = cast(Type) this;
         // `mutableOf` needs a mutable `this` only for caching
-        return cast(inout(Type)) unqualThis.mutableOf();
+        try
+            return cast(inout(Type)) unqualThis.mutableOf();
+        catch (Throwable) assert(0);
```
Comment (adamdruppe) [inline compiler/src/dmd/mtype.d:2397]: again

**Review by adamdruppe — CHANGES_REQUESTED** [review]

File: compiler/src/dmd/func.d (hunk 1 above)
Comment (WalterBright) [reply compiler/src/dmd/func.d:2037]: Throwable influences the `nothrow` attribute. Besides, this is just temporary scaffolding.

Also, assert's can throw an Exception, and wouldn't want to catch them here.

Comment (adamdruppe) [reply compiler/src/dmd/func.d:2037]: catching Exception satisfies nothrow. And asserts do not throw Exceptions and even fi they did, catch(Throwable) includes that anyway.

File: compiler/src/dmd/mtype.d (hunk 2 above)
Comment (WalterBright) [reply compiler/src/dmd/mtype.d:2397]: Putting it another way, `nothrow` is a promise that no `Throwables` are thrown. It is not a promise about `Exceptions`.

**Comment by WalterBright** [comment]
@adamdruppe you are right. I was wrong. I'll fix it.

**Comment by ibuclaw** [comment]
Did anyone actually try this with `-fno-exceptions`? If it fails I'll have to revert this.

**Pull request merged**